File: src/main/scala/ch_8_recursion/Q_3_Magic_Index.scala
package ch_8_recursion

class Q_3_Magic_Index {

  def find(array: Array[Int]): Int = {
    def find(start: Int, end: Int): Int = {
      val mid = (start + end) / 2
      if(mid == array(mid)) mid
      else if (mid < array(mid)) find(start, mid)
      else find(mid, end)
    }

    find(0, array.length)
  }
}

File: src/main/java/cci/recursion_dynamic/Q_3_MagicIndex.java
package cci.recursion_dynamic;

public class Q_3_MagicIndex
{
    public static int magicIndex(int[] arr)
    {
        return magicIndex(arr, 0, arr.length - 1);
    }

    private static int magicIndex(int[] arr, int start, int end)
    {
        if (start > end) {
            return -1;
        }

        int m = (start + end) / 2;

        if (arr[m] == m)
            return m;

        if (arr[m] > m)
            return magicIndex(arr, start, m);

        return magicIndex(arr, m, end);
    }

    public static void main(String[] args)
    {
        System.out.println(magicIndex(new int[]{
            -40, -20, -1, 1, 2, 3, 6, 7, 9, 12, 13
        }));
    }
}
